docs/discussion/Cosign Upgrade 2024.md

```diff
@@ -0,0 +1,468 @@
+      - ~~Implement a new cache provider for cache sharing between processes on same host~~
+        - ~~NOT trivial~~
+        - ~~Require a new http cache server and clients~~
+    - ~~Cache provider initialization in cosign plugin implementation~~
+    - ~~One advantage is that not only will cert + key sharing be solved but also oras store credentials can easily shared which is currently not possible for external plugins~~
+- Extend Inline Certificate Provider
+  - add new `type` field to spec: "key" or "certificate". If not provided, default to "certificate" for backwards compatability. 
+- Extend Azure Key Vault Certificate Provider
+  - use `GetKeys` API to fetch keys from configured key vault
+  - add new `keys` field to spec
+- Multitenancy considerations
+  - Need a separate keys map for namespaced and global certificate store keys
+
+### Option 2: Renaming CertificateStore to KeyManagementSystem
+- Same as Option 1 except we introduce a new CRD that will replace `CertificateStore` due to potential naming confusion
+- Is the existing name confusing enough to warrant changing to KeyManagementSystem?
+- How would we support CRD name change?
+  - Introduce new `KeyManagementSystem` CRD resource and then deprecate `CertificateStore`
+
+### Option 3: Introduce new CRD `KeyStore` separate from `CertificateStore`
+
+All of the same considerations as Option 1. It will be a completely new CRD with an inline and AKV provider equivalent.
+
+Pros: 
+- More intuitive naming: embedding 'key' specification in cert store might be confusing
+- No breaking changes or backwards compatability requirements
+Cons:
+- Yet one more resource type user must be concerned to configure
+- Providers will share almost the exact same implementation as certificate stores
+
+### How do we support CLI scenario?
+  - `CertificateStore` is only supported in K8s scenarios as CRD. There is no CLI equivalent.
+  - Ideally, `CertificateStore` should be supported at CLI too. However, even with CLI mode, a local directory for keys should be an option user can specify rather than just in-line and AKV.
+  - Cosign plugin will handle directory configuration and key reading within plugin implementation. This should be complimentary to any certs provided as plugin input.
+  - Should we have a separate local directory cert store provider that gets configured like other cert stores or should we scope it only to the cosign plugin?
+    - In K8s scenarios, it will not be an encouraged pattern to specify local directory. This is why implementing directory cert reading as a cert store provider may not make  sense.
+    - However, if a user decides to do a mixture of cert stores + local directory then they would need to specify in 2 different sections (cosign plugin config + cert store)
+    - **UPDATE 1/18/24**: We will mirror notation experience with a `verificationKeys` field in the cosign verifier config
+
+## Support Multiple keys for cosign verification
```

Comment:
TODO: make sure our docs for cosign + notation call out capabilities of mulit key + cert by default.